bot_core.js

```diff
@@ -125,6 +125,31 @@ const choose = async (context) => {
         reply_to_message_id: context.message.message_id,
       });
     }
+  } else if (fileName && fileName.endsWith(`.tdesktop-theme`)) {
+    bot.telegram.sendChatAction(context.message.chat.id, `upload_photo`);
+    const theme = await context.downloadFile(
+      context.message.document.file_id
+    );
+    const preview = await render({
+      theme,
+      name: fileName.replace(`.tdesktop-theme`, ``),
+      template: DESKTOP_TEMPLATE,
+    });
+    try {
+      await context.replyWithPhoto(
+        { source: preview },
+        {
+          reply_to_message_id: context.message.message_id,
+          caption: `Design by @voidrainbow`,
+        }
+      );
+    } catch (error) {
+      if (RESEND_ON_ERRORS.includes(error.name)) {
+        process.nextTick(sendPreview);
+      } else {
+        console.error(error);
+      }
+    }
```

Comment:
This function is now almost 100 lines long though it doesn't need to be this long.

First of all, one doesn't need to check twice if `fileName` exists. If it doesn't exist, there's nothing one can do and so has to return early from the function. After the return, `fileName` is guaranteed to exist.

Second, you can also extract the code that sends the theme, all that changes is `preview`, the caption and the keyboard.

So far you'll have:

```js
const fileName = ..;

if (!fileName) {
  return;
}

bot.telegram.sendChatAction(..);

let preview;
let caption;
let inlineKeyboard;

if (fileName.endsWith(`.attheme`)) {
  // ..
} else if (fileName.endsWith(`.tdesktop-theme`)) {
  // ..
}

const sendPreview = () => { .. };
sendPreview();
```

Further, we can do

```js
const extenstion = path.extname(fileName);
const name = path.basename(fileName, extension);
```

then we can replace `let preview` with `let theme; let template`, and in the if block we only fill `theme`, `template`, `caption` and `inlineKeyboard`. And then we'll be able to do this right before `sendPreview`:

```js
const preview = await render({ theme, name, template });
```

And in the end we'll have a pretty short function that's also more maintainable.